// bitcore-node-safe/lib/coinbaseToExcel/ExcelCount.js
var Excel = require('exceljs');
var mongo = require("mongoskin");
var uri ="mongodb://localhost:27017/coinbase";
var options = {
    server : {
        socketOptions : {
            socketTimeoutMS : 0,
            connectionTimeout : 0
        }
    }
};
function ExcelCount() {
    this.db = mongo.db(uri, options);
    this.db.bind("coinbase");

    this.workbook = new Excel.stream.xlsx.WorkbookWriter({
        filename: '/root/coinbaseCount.xlsx'
    });
    this.worksheet = this.workbook.addWorksheet('Sheet');
    this.worksheet.columns = [
        { header: '总额', key: 'sum' },
        { header: '地址', key: '_id' }
    ];
}

ExcelCount.prototype.generateExcelCountFile = function(option,callback){
    var self = this;
    self.db.coinbase.aggregate([{$match:option},{$group:{_id:"$address",sum:{$sum:"$value"}}}],function (err,array) {
        for(var i in array) {
            self.worksheet.addRow(array[i]).commit();
        }
        self.workbook.commit();
        callback();
    })
}

module.exports = ExcelCount;
