// bitcore-node-safe/lib/coinbaseToExcel/ExcelDetail.js
var Excel = require('exceljs');
var mongo = require("mongoskin");
var uri ="mongodb://localhost:27017/coinbase";
var options = {
    server : {
        socketOptions : {
            socketTimeoutMS : 0,
            connectionTimeout : 0
        }
    }
};

var ExcelDetail = function () {
    this.db = mongo.db(uri, options);
    this.db.bind("coinbase");
    this.workbook = new Excel.stream.xlsx.WorkbookWriter({
        filename: '/root/coinbaseList.xlsx'
    });
    this.worksheet = this.workbook.addWorksheet('Sheet');
    this.worksheet.columns = [
        { header: '奖励', key: 'value' },
        { header: '地址', key: 'address' },
        { header: '区块高度', key: 'height' },
        { header: '区块时间', key: 'time' }
    ];
}
ExcelDetail.prototype.generateExcelDetailFile = function (option,callback) {
    var self = this;
    self.db.coinbase.find(option).toArray(function(err,array){
// 开始添加数据
        for(var i in array) {
            array[i].time = timestampToTime(array[i].time);
            self.worksheet.addRow(array[i]).commit();
        }
        self.workbook.commit();
        callback();
    })
}

function timestampToTime(timestamp) {
    var date = new Date(timestamp * 1000);//时间戳为10位需*1000，时间戳为13位的话不需乘1000
    Y = date.getFullYear() + '-';
    M = (date.getMonth()+1 < 10 ? '0'+(date.getMonth()+1) : date.getMonth()+1) + '-';
    D = date.getDate() + ' ';
    h = date.getHours() + ':';
    m = date.getMinutes() + ':';
    s = date.getSeconds();
    return Y+M+D+h+m+s;
}

module.exports = ExcelDetail;
