// bitcore-node-safe/lib/coinbaseToExcel/RPCApi.js
var http = require('http');
var txId = "140d97d00cc8abb6855e5974d457c26a28060c2d9f24898953fdb580b58fb410";
var options={
    rpcuser:'rpcuser',
    rpcpassword:'rpcpass',
    host: '127.0.0.1',
    path: '/',
    method: 'POST',
    port: 8887,
    rejectUnauthorized: false,
    agent: undefined
};
var mycallback;
var req;
function  RPCApi() {
    req = http.request(options, function(res) {
        var buf='';
        res.on('data', function(data) {
            buf += data;
        });
        res.on('end', function() {
            var parsedBuf = JSON.parse(buf);
            mycallback(null,parsedBuf);
        });
    });
    req.on('error', function(e) {
        console.log(e);
    });
}

RPCApi.prototype.getInfo = function(callback){
    var request={"method":"getinfo","params":[],"id":50871};
    mycallback = callback;
    connectRPC(request);
}

RPCApi.prototype.getBlockHash = function(blcokHeigth,callback){
    var request={"method":"getblockhash","params":[],"id":50871};
    request["params"].push(blcokHeigth);
    mycallback = callback;
    connectRPC(request);
}

RPCApi.prototype.getBlockObject = function(blcokHash,callback){
    var request={"method":"getblock","params":[],"id":50871};
    request["params"].push(blcokHash);
    mycallback = callback;
    connectRPC(request);
}

RPCApi.prototype.getRawTransaction = function(txId,callback){
    var request={"method":"getrawtransaction","params":[],"id":50871};
    request["params"].push(txId);
    request["params"].push(1);
    mycallback = callback;
    connectRPC(request);
}

var connectRPC = function(option){
    var auth = new Buffer(options.rpcuser + ':' + options.rpcpassword).toString('base64');
    var requestData = JSON.stringify(option);
    req.setHeader('Content-Length', requestData.length);
    req.setHeader('Content-Type', 'application/json');
    req.setHeader('Authorization', 'Basic ' + auth);
    req.write(requestData);
    req.end();
}

module.exports = RPCApi;
