// bitcore-node-safe/lib/services/DBUtils/MysqlUtils.js
"use strict";

var mysql = require("mysql");

function MysqlUtils(mysqlConf) {
    this.connection = mysql.createConnection(mysqlConf);
    this.connection.connect();
}

MysqlUtils.prototype.subStrLastTowTurnUpperCase = function (str) {
    return str.substr(str.length - 2, str.length).toLocaleUpperCase()
};

// get data for bitcored start -------------------------------------------------------
/**
 * 查询地址锁定的总金额
 * 传入的参数，可以分页，排序(锁定金额和时间)
 * @param params
 * @param cb
 */
MysqlUtils.prototype.getAddressLockedTotal = function (params, cb) {
    var sql = "SELECT SUM(lockedVal) AS lockeVal, address, COUNT(id) AS lockCount " +
        "FROM lockedAddresses " +
        "WHERE unlockedHeight > ? " +
        "GROUP BY address " +
        "ORDER BY lockeVal  " + params.order +
        " LIMIT ?,10 ";
    var option = [params.unlockedHeight, params.start];
    this.sql(sql, option, function (err, result) {
        if (err) {
            console.log("get address locked total err " + JSON.stringify(err));
            cb(err);
        }
        cb(null, result);
    })
};

/**
 * 获取地址 资产锁定列表
 * @param params
 * @param cb
 */
MysqlUtils.prototype.getAssetLockedTotalForAssetName = function (params, cb) {
    var sql = "SELECT SUM(lockedVal) AS lockeVal, address, COUNT(id) AS lockCount " +
        "FROM lockedAssetAddress " +
        "WHERE unlockedHeight > ?  AND assetName = ?" +
        "GROUP BY address " +
        "ORDER BY lockeVal " + params.order +
        " LIMIT ?,10 ";
    var option = [params.unlockedHeight, params.assetName, params.start];
    this.sql(sql, option, function (err, result) {
        if (err) {
            console.log("get address locked total for asset name err " + JSON.stringify(err));
            cb(err);
        }
        cb(null, result)
    })
};

/**
 * 返回地址的总数
 * @param params 表名和条件
 * @param cb
 */
MysqlUtils.prototype.getAddressCount = function (params, cb) {
    var sql = "SELECT count(*) as count FROM(SELECT id FROM lockedAddresses WHERE unlockedHeight > ? GROUP BY address) a";
    var option = [params.unlockedHeight];
    if (params.assetName !== "safe") {
        sql = "SELECT count(*) as count FROM(SELECT id FROM lockedAssetAddress WHERE unlockedHeight > ? AND assetName = ? GROUP BY address) a";
        option.push(params.assetName);
    }
    this.sql(sql, option, cb);
};

/**
 * 根据页数返回锁定的output
 * @param params = {
 *  pageStart
 * }
 * @param cb
 */
MysqlUtils.prototype.getTransactionLockedTotal = function (params, cb) {
    var sql = "SELECT * FROM lockedAddresses " +
        " WHERE unlockedHeight > ? " + params.order +
        " LIMIT ?,10 ";
    var option = [params.unlockedHeight, params.start];
    this.sql(sql, option, function (err, result) {
        if (err) {
            console.log("get transaction locked total err " + JSON.stringify(err));
            cb(err);
        }
        cb(null, result);
    })
};

MysqlUtils.prototype.getTransactionLockedTotalForAssetName = function (params, cb) {
    var sql = "SELECT * FROM lockedAssetAddress WHERE unlockedHeight > ?  AND assetName = ? "
        + params.order + " LIMIT ?,10 ";
    var option = [params.unlockedHeight, params.assetName, params.start];

    this.sql(sql, option, function (err, result) {
        if (err) {
            console.log("get transaction locked total err " + JSON.stringify(err));
            cb(err);
        }
        cb(null, result);
    });
};

/**
 * 返回总数
 * @param params 表名和条件
 * @param cb
 */
MysqlUtils.prototype.getTransactionCount = function (params, cb) {
    var sql = "SELECT count(*) as count FROM lockedAddresses WHERE unlockedHeight > ?";
    var option = [params.unlockedHeight];
    if (params.assetName !== "safe") {
        sql = "SELECT count(*) as count FROM lockedAssetAddress WHERE unlockedHeight > ?  AND assetName = ?";
        option.push(params.assetName);
    }

    this.sql(sql, option, cb);
};

/**
 * 获取地址锁定详情
 * @param params
 * @param cb
 * @constructor
 */
MysqlUtils.prototype.LockedDetailByAddressDB = function (params, cb) {
    var sql = "SELECT txId,lockedVal,unlockedHeight,txTime FROM " + params.tableName +
        " WHERE unlockedHeight > ? AND address = ? " + params.where +
        "ORDER BY lockedVal " + params.order +
        " LIMIT ?,10";
    var option = [params.unlockedHeight, params.address, params.start];
    this.sql(sql, option, function (err, result) {
        if (err) {
            return cb(err)
        }
        cb(null, result);
    })
};

/**
 * 获取地址锁定的总数
 * @param params
 * @param cb
 * @constructor
 */
MysqlUtils.prototype.LockedDetailByAddressCountDB = function (params, cb) {
    var sql = "SELECT count(*) as count FROM " + params.tableName +
        " WHERE unlockedHeight > ? AND address = ? " + params.where;
    var option = [params.unlockedHeight, params.address];
    this.sql(sql, option, cb);
};

/**
 * 获取资产的锁定总金额
 * @param params
 * @param cb
 */
MysqlUtils.prototype.getTotalLockedAmountForAssetName = function (params, cb) {
    var sql = "SELECT SUM(lockedVal) AS totalLocked  FROM lockedAddresses WHERE unlockedHeight > ?;";
    var option = [params.unlockedHeight];
    if (params.assetName != "safe") {
        sql = "SELECT SUM(lockedVal) AS totalLocked FROM lockedAssetAddress WHERE unlockedHeight > ? AND assetName = ? ;";
        option = [params.unlockedHeight, params.assetName];
    }
    this.sql(sql, option, cb);
};

//  get data for bitcored end ------------------------------------------------

/**
 * 获取已经同步的区块高度
 * @param cb
 */
MysqlUtils.prototype.getAsyncBlockHeight = function (cb) {
    var sql = "select * from block";
    var option = [];
    this.sql(sql, option, function (err, result) {
        if (err) {
            console.log("get async block height err : " + JSON.stringify(err));
            cb(null);
        }
        cb(null, result[0]);
    })
};

/**
 * 更新已经同步的区块高度
 * @param height
 * @param cb
 */
MysqlUtils.prototype.updateAsyncBlockHeight = function (height, cb) {
    var sql = "update block set height = ?";
    var option = [height];
    var self = this;
    self.sql(sql, option, function (err, result) {
        if (err) {
            console.log("update async block height err  : " + JSON.stringify(err))
        }
        if (result.affectedRows == 0) {
            sql = "insert into block(height) value(?)";
            self.sql(sql, option, function (err, result) {
                if (err) {
                    console.log("insert async block height err  : " + JSON.stringify(err))
                }
                cb(null, result);
            })
        } else {
            cb(null, result);
        }
    })
};

/**
 * 添加safe锁定交易
 * @param out
 *      id 唯一标识
 *      txtime 交易时间
 *      address 接收地址
 *      txId 交易id
 *      lockedVal 锁定金额
 *      lockedHeight 锁定高度
 *      unlockedHeight 解锁高度
 *      lockedValInterest 利息
 * @param cb
 */
MysqlUtils.prototype.insertToSafe = function (out, cb) {
    var sql = "insert into lockedAddresses(id,txTime,address,txId,lockedVal,lockedHeight,unlockedHeight,lockedValInterest) value(?,?,?,?,?,?,?,?)";
    var option = [out._id, out.txTime, out.address, out.txId, out.lockedVal, out.lockedHeight, out.unlockedHeight, out.lockedValInterest];
    this.sql(sql, option, function (err, result) {
        if (err) {
            console.log("insert to safe locked data err :" + JSON.stringify(err));
            if (err.code != "ER_DUP_ENTRY") {
                return cb(err);
            }
        }
        cb(null, result);
    })
};

/**
 * 添加资产锁定交易
 * @param out
 *      id 唯一标识
 *      txtime 交易时间
 *      address 接收地址
 *      txId 交易id
 *      lockedVal 锁定金额
 *      lockedHeight 锁定高度
 *      unlockedHeight 解锁高度
 *      lockedValInterest 利息
 * @param cb
 */
MysqlUtils.prototype.insertToAssets = function (out, cb) {
    var sql = "insert into lockedAssetAddress(id,txTime,address,txId,lockedVal,lockedHeight,unlockedHeight,lockedValInterest,assetName) value(?,?,?,?,?,?,?,?,?)";
    var option = [out._id, out.txTime, out.address, out.txId, out.lockedVal, out.lockedHeight, out.unlockedHeight, out.lockedValInterest, out.assetName];
    this.sql(sql, option, function (err) {
        if (err) {
            console.log("insert to asset locked date err :" + JSON.stringify(err));
            return cb(err);
        }
        cb(null)
    })
};

/**
 * 添加asset信息
 * @param asset
 * @param cb
 */
MysqlUtils.prototype.insertAssetInfo = function (asset, cb) {
    var sql = "INSERT INTO asset VALUES(?,?,?)";
    var option = [asset.assetId, asset.name, asset.shortName];
    this.sql(sql, option, function (err, result) {
        if (err) {
            if (err.code != "ER_DUP_ENTRY") {
                return cb(err);
            }
        }
        cb(null, result);
    })
};

/**
 * 根据资产ID获取资产详情
 * @param name
 * @param cb
 */
MysqlUtils.prototype.getAssetNameByAssetId = function (name, cb) {
    var sql = "SELECT * FROM asset WHERE assetId = ?";
    this.sql(sql, [name], function (err, result) {
        if (err) {
            return cb(err);
        }
        cb(null, result);
    })
};

/**
 * 根据资产简称获取资产列表
 * @param name
 * @param cb
 */
MysqlUtils.prototype.getAssetNameByShortName = function (name, cb) {
    var sql = "SELECT assetShortName as name FROM asset ";
    if (name) {
        sql += "WHERE assetShortName like '%" + name + "%'";
    }
    this.sql(sql, [], function (err, result) {
        if (err) {
            return cb(err);
        }
        cb(null, result);
    })
};

/**
 * 按时间获取锁定列表
 * @param params
 * @param cb
 */
MysqlUtils.prototype.getAddressListForDate = function (params, cb) {
    var sql = "SELECT id as chainId,txTime,address,txId,lockedVal as amount,lockedHeight,unlockedHeight,lockedValInterest from lockedAddresses where txTime > ? and txTime < ? and lockedValInterest > 0";
    // var sql="SELECT * from lockedAddresses where txTime > ? and txTime < ?";
    var option = [params.startTime, params.endTime];
    this.sql(sql, option, function (err, result) {
        if (err) {
            return cb(err);
        }
        cb(null, result);
    })
};

/*** ADD BY ZY - 2021/09/28 针对钱包余额进行排名 */
/**
 * 更新/添加 账户资产余额
 * @param {*} params {
 *              amount:                     // 资产余额       
 *              latestTxHeight:             // 最后交易高度    
 *              latestUpdateHeight:         // 最后更新高度
 *              address:                    // 钱包地址        
 *            }
 * @param {*} callback 
 */
MysqlUtils.prototype.updateAddressBalance = function( params , callback ){
    var sql = "UPDATE balanceaddress SET balance = ?, locked = ?, latest_tx_height = ?, latest_update_height = ? WHERE address = ?";
    var option = [ params.balance , params.locked, params.latestTxHeight , params.latestUpdateHeight , params.address ];
    var self = this;
    self.sql( sql, option, function( err,result ){
        if (err) console.log("执行更新地址余额 [:balanceaddress] 失败  : " + JSON.stringify(err));
        if (result.affectedRows != 0){
            callback( err , result );
            return;
        }else{
            sql = "INSERT INTO balanceaddress( address , balance , locked, latest_tx_height,latest_update_height ) VALUES ( ?,?,?,?,? )";
            option = [ params.address , params.balance , params.locked, params.latestTxHeight, params.latestUpdateHeight ];
            self.sql(sql, option, function (err, result) {
                if (err) {
                    console.log("执行添加地址余额 [:BALANCEADDRESS] 失败  : " + JSON.stringify(err))
                }
                callback(null, result);
            });
        }
    } );
}

/**
 * 根据传入的 address , 查询对应的余额数据
 * @param {*} address 
 * @param {*} callback 
 */
MysqlUtils.prototype.getAddressBalance = function( address , callback ){
    var sql = "SELECT address AS address , balance AS balance , locked AS locked, latest_tx_height AS latestTxHeight,latest_update_height AS latestUpdateHeight FROM balanceaddress WHERE address = ?";
    var option = [ address ];
    var self = this;
    self.sql( sql , option , function( err , result ){
        if ( err ) console.log( "执行查询地址余额 [:balanceaddress] 失败 :" + JSON.stringify(err));
        callback( err , result );
    } )
}

/**
 * 统计地址余额
 * @param {*} callback 
 */
MysqlUtils.prototype.getTotalAddressBalance = function( callback ){
    var sql = "SELECT SUM(balance) AS totalAmount FROM balanceaddress";
    var option = [];
    var self = this;
    self.sql( sql , option , function( err , result ){
        if ( err ) console.log( "执行统计地址余额 [:balanceaddress] 失败 :" + JSON.stringify(err));
        callback( err , result );
    } )
}

/**
 * 查询账户资产余额排名
 * @param {*} params 
 * @param {*} callback 
 */
MysqlUtils.prototype.getBalanceRanking = function( params , callback ){
    var sql = "SELECT * FROM balanceaddress WHERE _del IS NULL ORDER BY balance DESC LIMIT ?,10";
    var option = [params.start];
    this.sql( sql, option, function( err , result ){
        if(err) console.log( "查询地址余额 [:BALANCEADDRESS] 错误 : " +  JSON.stringify(err));
        callback( null , result );
    } )
}
/**
 * 更新高度记录 , 此记录表用来追踪统计账户资产时的进度
 * @param {*} height 
 * @param {*} callback 
 */
MysqlUtils.prototype.updateBlockAmountStatisticRecord = function( height , callback ){
    var sql = "UPDATE block_balancestatistic SET height = ?";
    var option = [height];
    this.sql( sql, option, function( err , result ){
        if(err) console.log( "更新高度记录 [:BLOCK_BALANCESTATISTIC] 错误:" + JSON.stringify(err) )
        callback( null , result );
    } );
}

MysqlUtils.prototype.updateBlockAmountStatisticRecord2 = function( params , callback ){
    var sql = "UPDATE block_balancestatistic SET height = ? , total_address_amount = ? , total_locked_amount = ? , total_destory_amount = ? , total_flow_amount = ? , team_locked_amount = ? , masternode_count = ?";
    var option = [params.height,params.totalAddressAmount,params.totalLockedAmount,params.totalDestoryAmount,params.totalFlowAmount,params.teamLockedAmount,params.masternodeCount];
    this.sql( sql, option, function( err , result ){
        if(err) console.log( "更新高度记录 [:BLOCK_BALANCESTATISTIC] 错误:" + JSON.stringify(err) )
        callback( null , result );
    } );
}

/**
 * 获取追踪统计账户资产的高度记录
 */
MysqlUtils.prototype.getBlockAmountStatisticHeight = function( callback ){
    var sql = 'SELECT height FROM block_balancestatistic';
    var option = [];
    this.sql( sql, option, function( err , result ){
        if(err) console.log( "查询高度记录 [:BLOCK_BALANCESTATISTIC] 错误 : " +  JSON.stringify(err));
        callback( null , result );
    } )
}

MysqlUtils.prototype.getBlockAmountStatistic = function( callback ){
    var sql = 'SELECT height , total_amount AS totalAmount , total_address_amount AS totalAddressAmount , total_locked_amount AS totalLockedAmount, total_destory_amount AS totalDestoryAmount , total_flow_amount AS totalFlowAmount , team_locked_amount as teamLockedAmount ,masternode_count AS masternodeCount, masternode_unlock_tx_count AS masternodeUnlockTxCount,( select count(1) from balanceaddress ) address_count ,(select count(1) from balanceaddress where balance > 0) address_0_count FROM block_balancestatistic';
    var option = [];
    this.sql( sql, option, function( err , result ){
        if(err) console.log( "查询高度记录 [:BLOCK_BALANCESTATISTIC] 错误 : " +  JSON.stringify(err));
        callback( null , result );
    } )
}

/**
 * 从数据库中获取是否存在与当前高度解锁的钱包地址
 * @param {*} height 
 * @param {*} callback 
 */
MysqlUtils.prototype.getUnlockAddress = function( height , callback ){
    var sql = "select DISTINCT address from lockedAddresses where unlockedHeight = ?";
    var option = [height];
    this.sql( sql, option, function( err , result ){
        if(err) console.log( "获取指定高度解锁交易钱包地址 [:lockedaddresses] 错误 : " +  JSON.stringify(err));
        callback( null , result );
    } )
}

MysqlUtils.prototype.getStatistic = function( callback ){
    var sql = "select block_balancestatistic.* , ( select count(1) from balanceaddress ) address_count ,(select count(1) from balanceaddress where balance > 0) address_0_count from block_balancestatistic"
    var option = [];
    this.sql( sql, option, function( err , result ){
        callback( null , result );
    } );
}

MysqlUtils.prototype.getTop100 = function( callback ){
    var self = this;
    var top100Sql = "select address,balance from balanceaddress where _del is null order by balance desc limit 100";
    var count0Sql = "select count(1) address_0_count from balanceaddress where balance > 0";
    var option = [];
    var Result = {
        list:[],
        count:0
    }
    this.sql( top100Sql, option, function( err , result ){
        Result.list = result;
        self.sql( count0Sql, option, function( err , result ){
            Result.count = result[0].address_0_count;
            callback( null , Result )
        });
    });
}

MysqlUtils.prototype.getVpnAddress = function( callback ){
    var sql = "select address,vpn_id from address_seedvpn where type = 2";
    var option = [];
    this.sql( sql, option, function( err , result ){
        callback( null , result );
    } );
}
MysqlUtils.prototype.getSeedAddress = function( callback ){
    var sql = "select address from address_seedvpn where type = 1";
    var option = [];
    this.sql( sql, option, function( err , result ){
        callback( null , result );
    } );
}

/**
 * 获取锁仓数量为 1000 且对于指定高度已解锁的交易ID
 * @param {*} height 
 * @param {*} callback 
 */
MysqlUtils.prototype.getUnlockMasternodeTxList = function( height , callback ){
    var sql = "select id from lockedAddresses where lockedval = 1000 and unlockedheight <= ?"
    var option = [ height ];
    this.sql( sql, option, function( err , result ){
        callback( err , result );
    } );
}

MysqlUtils.prototype.updateMasternodeUnlockTxCount = function (count, cb) {
    var sql = "update block_balancestatistic set masternode_unlock_tx_count = ?";
    var option = [count];
    var self = this;
    self.sql(sql, option, function (err, result) {
        cb(err, result);
    })
};


/*** ---------------------------------------------------------------------------------------- */
MysqlUtils.prototype.sql = function (sql, option, callback) {
    var self = this;
    self.connection.query(sql, option, function (err, result) {
        if (err) {
            console.log("mysql Utils  err :" + err);
            return callback(err);
        }
        callback(null, result);
    })
};

module.exports = MysqlUtils;
