// bitcore-node-safe/lib/services/DBUtils/sendRequest.js
"use strict";

var http = require("http");
var https = require("https");

var callback;

function SendRequest(conf) {
    this.http = conf.http == "http" ? http : https;
    this.options = {
        hostname: conf.host,
        port: conf.port,
        path: conf.path,
        method: conf.method,
        json: true,
        rejectUnauthorized: conf.rejectUnauthorized,
        headers: {
            'Accept': 'application/json;version=2.0',
            'Content-Type': 'application/json'
        }
    }
}

SendRequest.prototype.send = function (params, cb) {
    params = JSON.stringify(params);
    callback = cb;
    this.init(function (req) {
        req.on('error', function (e) {
            console.log("---------------- send request error ----------------------");
            console.log(JSON.stringify(params));
            console.log("connection refuse");
        });
        req.write(params);
        req.end();
    });
};

SendRequest.prototype.init = function (cb) {
    var self = this;
    var req = this.http.request(self.options, function (res) {
        res.setEncoding('utf-8');
        var responseString = '';
        res.on('data', function (data) {
            responseString += data;
            callback(null, data);
        });
        res.on('end', function (res) {
            //这里接收的参数是字符串形式,需要格式化成json格式使用
        });
    });

    return cb(req);
};

module.exports = SendRequest;
